pkg/docker/docker.go

```diff
@@ -768,19 +946,52 @@ func (d *stiDocker) RunContainer(opts RunContainerOptions) error {
 		os.Exit(2)
 	}
 	return interrupt.New(dumpStack, removeContainer).Run(func() error {
-		// Attach to the container.
+		// Attach to the container on go thread (different than with go-dockerclient, since it provided a non-blocking attach which we don't seem to have with k8s/engine-api)
+
+		// Attach to  the container on go thread to mimic blocking behavior we had with go-dockerclient (k8s wrapper blocks); then use borrowed code
+		// from k8s to dump logs via return
+		// still preserve the flow of attaching before starting to handle various timing issues encountered in the past, as well as allow for --run option
 		glog.V(2).Infof("Attaching to container %q ...", containerName)
-		attachOpts := opts.asDockerAttachToContainerOptions()
-		attachOpts.Container = container.ID
-		if _, err = d.client.AttachToContainerNonBlocking(attachOpts); err != nil {
-			glog.V(0).Infof("error: Unable to attach to container %q with options %+v: %v", containerName, attachOpts, err)
-			return err
+		errorChannel := make(chan error)
+		timeoutTimer := time.NewTimer(DefaultDockerTimeout)
+		var attachLoggingError error
+		go func() {
+			ctx, cancel := getDefaultContext(DefaultDockerTimeout)
+			defer cancel()
+			resp, err := d.client.ContainerAttach(ctx, container.ID, opts.asDockerAttachToContainerOptions())
+			errorChannel <- err
+			if err != nil {
+				glog.V(0).Infof("error: Unable to attach to container %q: %v", containerName, err)
+				return
+			}
+			defer resp.Close()
+			sopts := opts.asDockerAttachToStreamOptions()
+			attachLoggingError = d.holdHijackedConnection(sopts.RawTerminal, sopts.InputStream, sopts.OutputStream, sopts.ErrorStream, resp)
```

Comment:
no more so than with go-dockerclient 's non-blocking version of attach; the processing of stdout/stdin occurs in go routines there as well, in a similar fashion, after the "attach http post" has returned.